### PR TITLE
Context added TSA

### PR DESCRIPTION
**Pull request by kitaisreal**
Follow-up to #55121

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 6d08e289d6c40745906ecef73e16dd581c10cc25 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55278/6d08e289d6c40745906ecef73e16dd581c10cc25/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6434524260">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55278/6d08e289d6c40745906ecef73e16dd581c10cc25/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by rschu1ze** [comment]
The direction is good but TBH the `SharedMutexHelper` / `ContextSharedMutex` / `SharedLockGuard` machine seem overengineered. As far as I see, it addresses two problems:
1. Only the non-relockable locker classes in libcxx (`std::lock_guard` and `std::scoped_lock`) have TSA annotations, the relockable ones (`std::unique_lock`, `std::shared_lock`) lack them.
2. Additional code (update profile events) should run when the lock is aquired/released. As none of the locker classes in libcxx allows that, a custom implementation is needed.

I believe that 2. is quite a special case throughout the codebase, and that a non-generic (= `Context`-specific) class would do the job. I also think that the problem can be solved with small wrappers around `std::unique_lock` and `std::shared_lock` which behave like non-relockable locker classes and run some code in their ctors/dtors. Let me try to implement this (in a separate PR).

**Comment by kitaisreal** [comment]
> The direction is good but TBH the `SharedMutexHelper` / `ContextSharedMutex` / `SharedLockGuard` machine seem overengineered. As far as I see, it addresses two problems:
> 
> 1. Only the non-relockable locker classes in libcxx (`std::lock_guard` and `std::scoped_lock`) have TSA annotations, the relockable ones (`std::unique_lock`, `std::shared_lock`) lack them.
> 2. Additional code (update profile events) should run when the lock is aquired/released. As none of the locker classes in libcxx allows that, a custom implementation is needed.
> 
> I believe that 2. is quite a special case throughout the codebase, and that a non-generic (= `Context`-specific) class would do the job. I also think that the problem can be solved with small wrappers around `std::unique_lock` and `std::shared_lock` which behave like non-relockable locker classes and run some code in their ctors/dtors. Let me try to implement this (in a separate PR).

1. It does not do the same logic at all https://github.com/ClickHouse/ClickHouse/pull/55573/files#diff-c7c4cea868f661341c1e9866836dc34c1c88723f9f33b4e09db530c2ea074036R214. If we take mutex in `LockGuard` constructor and start to record time, and in destructor try to add elapsed time to `ProfileEvents`. It is not the same as implemented in my pull request, we need to record time that we actually waited for mutex to be acquired.
2. std::unique_lock, std::shared_lock provide additional overhead that is not needed at all, because your class now is movable and for TSA it is harder to analyze such classes.
3. SharedMutexHelper is needed in a lot of other places in codebase when we want to add custom logic for `lock` and `unlock` methods.

**Comment by kitaisreal** [comment]
Additionally:
4. We should add custom logic to mutex not to `LockGuards` because we actually can lock mutex without this special guard and that way we does not get custom logic to run.
5. `SharedLockGuard` is also needed in a lot of places in codebase where we use `shared_mutex` and want to add TSA.

**Comment by rschu1ze** [comment]
After checking your PR in more detail, I think it is okay to merge (after incorporating the feedback).

>     1. It does not do the same logic at all https://github.com/ClickHouse/ClickHouse/pull/55573/files#diff-c7c4cea868f661341c1e9866836dc34c1c88723f9f33b4e09db530c2ea074036R214. If we take mutex in `LockGuard` constructor and start to record time, and in destructor try to add elapsed time to `ProfileEvents`. It is not the same as implemented in my pull request, we need to record time that we actually waited for mutex to be acquired.

True, that was a glitch on my end.

>     2. std::unique_lock, std::shared_lock provide additional overhead that is not needed at all, because your class now is movable and for TSA it is harder to analyze such classes.

In my PR, TSA doesn't care what is inside both new lock classes. It only looks at the interface of the lock class (which is the same as in your PR - the constructor and destructor with annotation). There was a little bit of extra overhead but I don't think it would be significant.
 
>     3. SharedMutexHelper is needed in a lot of other places in codebase when we want to add custom logic for `lock` and `unlock` methods.

Not sure if there are "a lot" of other places but it is a good to make custom logic around `lock`/`unlock` possible.

> 4. We should add custom logic to mutex not to LockGuards because we actually can lock mutex without this special guard and that way we does not get custom logic to run.

That was where my PR took a different approach (to reduce complexity) but you are right that it also allowed to evade the custom logic.

> 5. SharedLockGuard is also needed in a lot of places in codebase where we use shared_mutex and want to add TSA.

I added a comment about SharedLockGuard into the PR.

**Comment by rschu1ze** [comment]
I checked the PR two days ago but somehow forgot to publish my comments. Adding them now - you addressed some of the comments already.

File: src/Common/SharedLockGuard.h [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <base/types.h>
+#include <base/defines.h>
+
+namespace DB
+{
+
+template <typename Mutex>
+class TSA_SCOPED_LOCKABLE SharedLockGuard
```
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Common/SharedLockGuard.h:11]: (resolved in the meantime)

File: src/Common/SharedLockGuard.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#pragma once
+
+#include <base/types.h>
+#include <base/defines.h>
+
+namespace DB
+{
+
+template <typename Mutex>
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/SharedLockGuard.h:10]: Updated.

File: src/Common/MutexUtils.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
+#pragma once
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/MutexUtils.h:1]: Updated.

File: src/Common/MutexUtils.h [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+#include <base/types.h>
+#include <base/defines.h>
+#include <Common/SharedMutex.h>
+
+namespace DB
+{
+
+template <typename Derived, typename MutexType = SharedMutex>
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply src/Common/MutexUtils.h:10]: Updated.

File: base/base/defines.h [hunk 5]
```diff
@@ -150,6 +150,7 @@
 #    define TSA_ACQUIRE_SHARED(...) __attribute__((acquire_shared_capability(__VA_ARGS__)))          /// function acquires a shared capability, but does not release it
 #    define TSA_TRY_ACQUIRE_SHARED(...) __attribute__((try_acquire_shared_capability(__VA_ARGS__)))  /// function tries to acquire a shared capability and returns a boolean value indicating success or failure
 #    define TSA_RELEASE_SHARED(...) __attribute__((release_shared_capability(__VA_ARGS__)))          /// function releases the given shared capability
+#    define TSA_SCOPED_LOCKABLE __attribute__((scoped_lockable)) /// object of a class has scoped lockable capability
```
[the remark this answers is not in the document]
Comment (kitaisreal) [reply base/base/defines.h:153]: Updated.

**Pull request merged**